edb/server/pgrust/src/conn_string.rs

```diff
@@ -0,0 +1,978 @@
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+pub enum Host {
+    Hostname(String, u16),
+    IP(IpAddr, u16, Option<String>),
+    Path(String, u16),
+    Abstract(String, u16),
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+pub enum Password {
+    /// The password is unspecified and should be read from the user's default
+    /// passfile if it exists.
+    #[default]
+    Unspecified,
+    /// The password was specified.
+    Specified(String),
+    /// The passfile is specified.
+    Passfile(PathBuf),
+}
+
+#[derive(Serialize)]
+pub enum PasswordWarning {
+    NotFile(PathBuf),
+    NotExists(PathBuf),
+    NotAccessible(PathBuf),
+    Permissions(PathBuf, u32),
+}
+
+impl std::fmt::Display for PasswordWarning {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PasswordWarning::NotFile(path) => write!(f, "Password file {path:?} is not a plain file"),
+            PasswordWarning::NotExists(path) => write!(f, "Password file {path:?} does not exist"),
+            PasswordWarning::NotAccessible(path) => write!(f, "Password file {path:?} is not accessible"),
+            PasswordWarning::Permissions(path, mode) => write!(f, "Password file {path:?} has group or world access ({mode:o}); permissions should be u=rw (0600) or less"),
+        }
+    }
+}
+
+#[cfg(windows)]
+const PGPASSFILE: &str = "pgpass.conf";
+#[cfg(not(windows))]
+const PGPASSFILE: &str = ".pgpass";
+
+impl Password {
+    pub fn password(&self) -> Option<&str> {
+        match self {
+            Password::Specified(password) => Some(password),
+            _ => None,
+        }
+    }
+
+    /// Attempt to resolve a password against the given homedir.
+    pub fn resolve(
+        &mut self,
+        home: &Path,
+        hosts: &[Host],
+        database: &str,
+        user: &str,
+    ) -> Result<Option<PasswordWarning>, std::io::Error> {
+        let passfile = match self {
+            Password::Unspecified => {
+                let passfile = home.join(PGPASSFILE);
+                // Don't warn about implicit missing or inaccessible files
+                if !matches!(passfile.try_exists(), Ok(true)) {
+                    *self = Password::Unspecified;
+                    return Ok(None);
+                }
+                if !passfile.is_file() {
+                    *self = Password::Unspecified;
+                    return Ok(None);
+                }
+                passfile
+            }
+            Password::Specified(_) => return Ok(None),
+            Password::Passfile(passfile) => {
+                let passfile = passfile.clone();
+                if matches!(passfile.try_exists(), Ok(false)) {
+                    *self = Password::Unspecified;
+                    return Ok(Some(PasswordWarning::NotExists(passfile)));
+                }
+                if passfile.exists() && !passfile.is_file() {
```

Comment:
`exists()` duplicates with the `try_exists()` above?

Reply:
It's not quite the same -- `try_exists` will return `Ok(...)` if and only if the file can be guaranteed to exist or not exist, while `exists()` returns `false` if the file definitely doesn't exist OR there was an error. This is somewhat awkward code that just preserves the existing error functionality so the tests pass with the correct failures.